### DISCUSSION TITLE
Generated types for a consistent DAO layer

### DISCUSSION THREAD

**Discussion by raymondKelly** [prisma/prisma#21013] [Q&A]
### Summary
I would like to have generated types that include no client query structure and instead are an exact typescript type match for the schema.prisma file so that I can build DRY functions for querying the database that does not require other services to have knowledge of the underlying client. 

### My Setup
I currently have the use case in my monorepo which is structured as follows

Mono
------> Product A
------------------> Services
----------------------------> DTOs
------------------> Database
----------------------------> DAOs
----------------------------> schema.prisma
----------------------------> Client

Inside my database project, I essentially have the database schema, the generated client and a DAOs to allow me to access the database in a consistent pattern across my services. The services should no how to structure the data before putting it into the database but should have no knowledge of the underlying client being used or how the query needs to be structured, instead my DRY functions should do that.

An example of my schema is as follows
```prisma
schema.prisma

model User {
    id                      String            @id @default(uuid())
    createdAt        DateTime       @default(now())
    modifiedAt       DateTime      @default(now()) @updatedAt
    name                String?            
    profile              Profile?
}

model Profile {
    id                      String            @id @default(uuid())
    createdAt        DateTime       @default(now())
    modifiedAt       DateTime      @default(now()) @updatedAt
    user                  User?            @relation(fields: [userId], references: [id], onDelete: Cascade)
    userId               String?          @unique          
    photos              Photo[]
}

model Photo {
    id                      String            @id @default(uuid())
    createdAt        DateTime       @default(now())
    modifiedAt       DateTime      @default(now()) @updatedAt
    profile               Profile?          @relation(fields: [profileId], references: [id], onDelete: Cascade)
    profileId            String?                   
}
```

```ts
user-dao.ts

  createUser = async (user: TYPE???) => {
  // I would like to build the query here to create the data. Iterate over the relations of the user and create those too(separate queries due to the bug with createMany)

  }
```

```ts
test-service.ts

import { createUser } from '../user-dao.ts';

const userCreationApi = () => {
 return createUser({id: '123', createdAt: new Date(), modifiedAt: new Date()})
}

```
### Problem
I can not get any types that keep the types solely to the schema which do not include Prisma Client types such as CreateMany

For example:

I have used 
```
type PostCreateBody = Prisma.Args<typeof client.user, 'create'>['data'];
```
However, this has the problem that PostCreateBody.profile does not have the schema for the profile and instead has `create`, `connect` and `createMany` which will not work as that would result in the services requiring knowledge of the underlying client and would essentially be the one then building the query.

I have also tried
```ts
type TestType = Prisma.UserGetPayload<{
  include: {
    profile: {
      include: {
        photos: true
      }
    }
  };
}>;
```
However, this then has the problem that everything on the type becomes a required field and I lose the optional `?` piece of the type which will not work as in this example I should be able to create a User with no name as that is optional but compile time fails as it is required using this type. 

Any help or guidance would be really appreciated.

**Comment by ludralph** [comment]
Hi @raymondKelly 👋 

Thank you for raising this question.

Our recommended approach for your use case is documented [here](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#problem-using-variations-of-the-generated-model-type).

> However, this then has the problem that everything on the type becomes a required field and I lose the optional ? piece of the type

Ideally, this should not be the case. If you have defined a field as optional in your schema file, the generated type should also have that field as optional. Could you provide a small reproduction so that we can take a look?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
